packages/hardware-ledger/src/LedgerKeyAgent.ts

```diff
@@ -148,98 +165,207 @@ export class LedgerKeyAgent extends KeyAgentBase {
+        connection.nodeHidDevicePath === nodeHidDevicePath &&
+        connection.device === device
     );
+    if (!matchingConnectionData) return null;
+
+    try {
+      await this.testConnection(matchingConnectionData.deviceConnection);
+    } catch (error) {
+      if (error instanceof errors.TransportError && error.message.includes(CARDANO_APP_CONNECTION_ERROR_MESSAGE)) {
+        this.deviceConnections = this.deviceConnections.filter(
+          (connectionData) => connectionData !== matchingConnectionData
+        );
+        return null;
+      }
+
+      throw error;
+    }
+
+    return matchingConnectionData.deviceConnection;
   }
 
   /**
    * @throws TransportError
    */
-  static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
+  private static async getHidDeviceList(communicationType: CommunicationType): Promise<string[]> {
     try {
-      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebHID.list();
+      return communicationType === CommunicationType.Node ? TransportNodeHid.list() : TransportWebUSB.list();
     } catch (error) {
       throw new errors.TransportError('Cannot fetch device list', error);
     }
   }
 
+  private static async closeTransport(transport: LedgerTransportType) {
+    try {
+      await transport.close();
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error('Failed to close the transport', transport, error);
+    }
+  }
+
+  private static attachDisconnectionCleanupHandler(transport: LedgerTransportType) {
+    const onDisconnect = () => {
+      transport.off('disconnect', onDisconnect);
+      this.deviceConnections = this.deviceConnections.filter(
+        ({ deviceConnection }) => deviceConnection.transport !== transport
+      );
+      void this.closeTransport(transport);
+    };
+    transport.on('disconnect', onDisconnect);
+  }
+
+  /**
+   * @throws TransportError
+   */
+  private static async openTransportForDevice({ communicationType, device }: OpenTransportForDeviceParams) {
+    let transport: LedgerTransportType;
+    try {
+      if (communicationType === CommunicationType.Node && device instanceof HID) {
+        transport = new TransportNodeHid(device);
+      } else if (communicationType === CommunicationType.Web && device instanceof USBDevice) {
+        transport = await TransportWebUSB.open(device);
+      } else {
+        throw new errors.TransportError(`Invalid device object provided for communication type ${communicationType}`);
+      }
+    } catch (error) {
+      throw new errors.TransportError('Failed to open a transport for a given device', error);
+    }
+    this.attachDisconnectionCleanupHandler(transport);
+    return transport;
+  }
+
   /**
    * @throws TransportError
    */
   static async createTransport({
     communicationType,
-    activeTransport,
-    devicePath = ''
+    nodeHidDevicePath = ''
```

Comment:
Breaking change here. I removed the `activeTransport` from params as I didn't find any usage. Is that ok? Are there any consumers of it that I'm not aware of?

Reply:
> Breaking change here. I removed the activeTransport from params as I didn't find any usage. Is that ok?
It might be breaking `tsc` (I think it depends on tsconfig) but is not a breaking change in js, so it's very minor and I wouldn't consider it breaking

Reply:
I meant a breaking change on an API level. If there was someone using this method we would need to release this change under a new major version, to avoid breaking his app.

Reply:
> I meant a breaking change on an API level

Me too, we are removing a property, not adding it, so if you have:
```ts
createTransport({ nodeHidDevicePath, ... })
```

It would still work after updating this package in most cases. For typescript users tsc might give you an error, but I believe not always (depends on tsconfig)